sdk/src/test/kotlin/com/pulumi/kotlin/options/ResourceOptionsTest.kt

```diff
@@ -0,0 +1,545 @@
+
+        // when
+        val opts = opts {
+            resourceTransformations(
+                { transformation ->
+                    transformationResult {
+                        args(transformation.args())
+                        options(transformation.options())
+                    }
+                },
+            )
+        }
+
+        // then
+        assertResourceTransformationResultEquals(
+            expectedResourceTransformationResult,
+            opts.resourceTransformations.orEmpty(),
+            javaArgs,
+        )
+    }
+
+    @Test
+    fun `option resourceTransformations should be properly set with list`() = runBlocking {
+        // given
+        val javaResourceMock = mockk<JavaResource>()
+        val javaResourceArgsMock = mockk<JavaResourceArgs>()
+        val javaResourceOptionsMock = mockk<JavaResourceOptions>()
+        val javaArgs = JavaResourceTransformationArgs(
+            javaResourceMock,
+            javaResourceArgsMock,
+            javaResourceOptionsMock,
+        )
+
+        val expectedResourceTransformationResult =
+            JavaResourceTransformationResult(javaResourceArgsMock, javaResourceOptionsMock)
+
+        val resourceTransformation = ResourceTransformation { transformation ->
+            transformationResult {
+                args(transformation.args())
+                options(transformation.options())
+            }
+        }
+
+        // when
+        val opts = opts {
+            resourceTransformations(listOf(resourceTransformation))
+        }
+
+        // then
+        assertResourceTransformationResultEquals(
+            expectedResourceTransformationResult,
+            opts.resourceTransformations.orEmpty(),
+            javaArgs,
+        )
+    }
+
+    @Test
+    fun `option retainOnDelete should be properly set`() = runBlocking {
+        // when
+        val opts = opts {
+            retainOnDelete(true)
+        }
+
+        // then
+        assertTrue(opts.retainOnDelete, "retainOnDelete")
+    }
+
+    @Test
+    fun `option urn should be properly set`() = runBlocking {
+        // when
+        val opts = opts {
+            urn("urn:pulumi:production::acmecorp::custom:resources:Resource:s3/bucket:Bucket::my-bucket")
+        }
+
+        // then
+        assertEquals(
+            "urn:pulumi:production::acmecorp::custom:resources:Resource:s3/bucket:Bucket::my-bucket",
+            opts.urn,
+            "opts.urn",
+        )
+    }
+
+    @Test
+    fun `option version should be properly set`() = runBlocking {
+        // when
+        val opts = opts {
+            version("1.0.0")
+        }
+
+        // then
+        assertEquals("1.0.0", opts.version, "opts.version")
+    }
+
+    @Test
+    abstract fun `options should be properly merged to existing options`()
+
+    @Test
+    abstract fun `empty options should contain only nulls or type defaults`()
```

Comment:
Maybe it would be better to just add these methods to subclasses without mentioning them here? 

Abstract methods in tests can be a bit surprising.